pkg/errctx/context.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errctx
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/errno"
+	"github.com/pingcap/tidb/pkg/util/intest"
+)
+
+// Level defines the behavior for each error
+type Level uint8
+
+const (
+	// LevelError means the error will be returned
+	LevelError Level = iota
+	// LevelWarn means it will be regarded as a warning
+	LevelWarn
+	// LevelIgnore means the error will be ignored
+	LevelIgnore
+)
+
+// Context defines how to handle an error
+type Context struct {
+	levelMap        [2]Level
+	appendWarningFn func(err error)
+}
+
+// WithStrictErrGroupLevel makes the context to return the error directly for any kinds of errors.
+func (ctx *Context) WithStrictErrGroupLevel() Context {
+	newCtx := Context{
+		appendWarningFn: ctx.appendWarningFn,
+	}
+
+	return newCtx
+}
+
+// WithErrGroupLevel sets a `Level` for an `ErrGroup`
+func (ctx *Context) WithErrGroupLevel(eg ErrGroup, l Level) Context {
+	newCtx := Context{
+		levelMap:        ctx.levelMap,
+		appendWarningFn: ctx.appendWarningFn,
+	}
+	newCtx.levelMap[eg] = l
+
+	return newCtx
+}
+
+// AppendWarning appends the error to warning. If the inner `appendWarningFn` is nil, do nothing.
+func (ctx *Context) AppendWarning(err error) {
```

Comment:
Is it possible to make `AppendWarning` private? I think it is not expected to be called directly by other code.